tests/lib/openstack_query/managers/test_query_manager.py

```diff
@@ -86,32 +90,62 @@ def test_build_and_run_query_with_runner_params(
             ("with both", MOCKED_PRESET_DETAILS, MOCKED_OUTPUT_DETAILS),
         ]
     )
-    @patch("openstack_query.managers.query_manager.QueryManager._populate_query")
-    @patch("openstack_query.managers.query_manager.QueryManager._get_query_output")
+    @patch("openstack_query.managers.query_manager.QueryOutputDetails")
     def test_build_and_run_query(
         self,
         _,
         mock_preset_details,
         mock_output_details,
-        mock_get_query_output,
-        mock_populate_query,
+        mock_query_output_details,
     ):
         """
         Tests that _build_and_run_query method functions expectedly
         Sets up a QueryResource object and runs a given query with appropriate inputs (no runner params).
         Returns query result
         """
-        res = self.instance._build_and_run_query(
-            mock_output_details, mock_preset_details
-        )
 
-        mock_populate_query.assert_called_once_with(
-            preset_details=mock_preset_details,
-            properties_to_select=mock_output_details.properties_to_select,
-        )
+        if not mock_output_details:
+            # just assume valid output details is made from calling from_kwargs
+            mock_query_output_details.from_kwargs.return_value = MOCKED_OUTPUT_DETAILS
+
+        with patch(
+            "openstack_query.managers.query_manager.QueryManager._populate_output_params"
+        ) as mock_populate_output_params:
+            with patch(
+                "openstack_query.managers.query_manager.QueryManager._get_query_output"
+            ) as mock_get_query_output:
+                res = self.instance._build_and_run_query(
+                    preset_details=mock_preset_details,
+                    output_details=mock_output_details,
+                )
+                self.assertEqual(res, mock_get_query_output.return_value)
+                mock_get_query_output.assert_called_once_with(
+                    mock_output_details.output_type
+                )
+
+                if not mock_output_details:
+                    mock_query_output_details.from_kwargs.assert_called_once_with(
+                        self.prop_cls
+                    )
+                    mock_populate_output_params.assert_called_once_with(
+                        output_details=mock_query_output_details.from_kwargs.return_value
+                    )
+                else:
+                    mock_query_output_details.from_kwargs.assert_not_called()
+                    mock_populate_output_params.assert_called_once_with(
+                        output_details=mock_output_details
+                    )
+
+        if mock_preset_details:
+            self.query.where.assert_called_once_with(
+                preset=MOCKED_PRESET_DETAILS.preset,
+                prop=MOCKED_PRESET_DETAILS.prop,
+                **MOCKED_PRESET_DETAILS.args,
+            )
+        else:
+            self.query.where.assert_not_called()
```

Comment:
If you see an if - else on a unit test then it's a sign you've got two test-cases smashed into one